katsdpcontroller/test/test_product_config.py

```diff
@@ -241,6 +245,35 @@ def test_from_config(self) -> None:
         assert_equal(cam_http.url, yarl.URL('http://test.invalid'))
 
 
+class TestSimDigRawAntennaVoltageStream:
+    """Test :class:`~.SimDigRawAntennaVoltageStream`."""
+
+    def setup(self):
+        self.config: Dict[str, Any] = {
+            'type': 'sim.dig.raw_antenna_voltage',
+            'adc_sample_rate': 1712000000.0,
+            'centre_frequency': 1284000000.0,
+            'band': 'l',
+            'antenna': _M000.description
+        }
+
+    def test_from_config(self) -> None:
+        dig = SimDigRawAntennaVoltageStream.from_config(
+            Options(), 'm000h', self.config, [], {}
+        )
+        assert_equal(dig.adc_sample_rate, 1712000000.0)
+        assert_equal(dig.centre_frequency, 1284000000.0)
```

Comment:
Perhaps my understanding of the unit-testing philosophy is wrong here.
Would it not be better to use references to `self.config.adc_sample_rate` for instance? Hardcoding the number seems prone to problems if you change one but forget to change the other.

Reply:
You're right, I am violating the don't-repeat-yourself principle. I've pushed some changes. I tend to prefer putting explicit values in rather than formulae because I've been burned before by having the formula be incorrect, whereas if I evaluate the formula by hand I can look at the result and say "that's not right" (once at an old job it was due to an `int` being cast to a C 32-bit `float` and losing precision in the process, which you couldn't immediately see by eye-balling the code). But in these cases the value I need is right there in the dictionary, no formula required.

Reply:
Changes look good to me.

But you do also make a point regarding evaluating test cases "by hand" so to speak. I'd just want stuff to be reviewed very very carefully if we do that. I'm good at making silly errors in by-hand calculations.

Reply:
As long as it doesn't exactly match a mistake in the calculation coded in the implementation, the unit test will tell you :-)